xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -431,18 +432,46 @@ AllocateDestinationBuffer(
     const Shape& on_host_shape, PjRtDevice* device,
     LocalDeviceState* local_device, se::Stream* copy_stream,
     bool is_uninitialized_create, PjRtStreamExecutorClient* client,
-    std::shared_ptr<BufferSequencingEvent> definition_event) {
+    std::shared_ptr<BufferSequencingEvent> definition_event,
+    PjRtMemorySpace* memory_space) {
   if (on_host_shape.IsTuple() && on_host_shape.tuple_shapes_size() == 0) {
     return InvalidArgument("Can't make a buffer from an empty tuple");
   }
 
+  PjRtMemorySpace* default_memory_space =
+      device->default_memory_space().value_or(nullptr);
+  if (!memory_space) {
+    memory_space = default_memory_space;
+  }
+  bool is_pinned_host_memory =
+      memory_space && (memory_space->kind() == PinnedHostMemorySpace::kKind);
+  // Only allow pinned host memory or device memory.
+  if (memory_space != default_memory_space && !is_pinned_host_memory) {
+    return InvalidArgument("Buffer allocation: invalid memory space");
+  }
+
   auto* se_client = tensorflow::down_cast<PjRtStreamExecutorClient*>(client);
   TransferManager* transfer_manager =
       se_client->client()->backend().transfer_manager();
-  TF_ASSIGN_OR_RETURN(ScopedShapedBuffer dst_buffer,
-                      transfer_manager->AllocateScopedShapedBuffer(
-                          on_host_shape, se_client->allocator(),
-                          local_device->local_device_id().value()));
+
+  // Communicate the desired memory space to the allocator via the shape
+  // callback.
+  auto memory_space_shape_fn = [is_pinned_host_memory,
+                                transfer_manager](const Shape& shape) {
+    Shape result = shape;
+    if (is_pinned_host_memory) {
+      result.mutable_layout()->set_memory_space(Layout::kHostMemorySpace);
+    } else {
+      result = transfer_manager->HostShapeToDeviceShape(result);
+    }
+    return result;
+  };
+
+  TF_ASSIGN_OR_RETURN(
+      ScopedShapedBuffer dst_buffer,
+      transfer_manager->AllocateScopedShapedBuffer(
+          on_host_shape, se_client->allocator(),
+          local_device->local_device_id().value(), memory_space_shape_fn));
```

Comment:
do you need to update https://github.com/openxla/xla/blob/b700711f832e87bcc2c3960f879e287f98b9737b/xla/stream_executor/stream_executor.h#L370-L378 as well? it looks like memory_limit_bytes check does not apply to kHostMemorySpace?

Reply:
I am not sure if the limit applies or not. Arguably the change to checking the limit is independent of this patch - the StreamExecutor::AllocateArray method is already getting the memory space as argument, so it should do the appropriate checking if that's important.

How would you feel about not updating the AllocateArray method in this patch?

Reply:
I was thinking skip this check if the memory_space is kHostMemorySpace? Current memory_limit_bytes is only for device memory. It is possible that this check will fail at `static_cast<int64_t>(bytes) > memory_limit_bytes`, however, the bytes will be allocated from host memory and should not matter if it is larger than memory_limit_bytes. What do you think? 

@ezhulenev Any suggestion?

Reply:
I don't disagree, I just think that should be a different PR. Would it help if I write such a PR now?

Reply:
A different PR sounds good. Sorry I misunderstood the earlier reply.

Reply:
Just to follow up here, it looks like the host memory allocator bypasses the AllocateArray method. Instead of AllocateArray, it call HostMemoryAllocate, see [here](https://github.com/openxla/xla/blob/5f4c1ebb3ef1b141e0f47452867c8c8dbc066983/xla/stream_executor/integrations/device_host_allocator.h#L59). 

As a result, I believe there is no need to worry about that limit check.